File: py110/small_problems_110_119/medium1/3.py
# https://launchschool.com/exercises/89604c6e

'''
Questions:
1. 


input: integer
output: rotated integer
rules:

length = 6 
rotations = 6

while loop

735291 # 7
0 1 2 3 4 5

slice[:0] + slice[0 + 1:] + slice[0]

352917 # kept first digit
slice[:1] + slice[1 + 1:] + slice[1]

329175 # keep first 2
321759 # keep 3
321597 # keep 4
321579

pseudocode:
1. convert input to string
2. Keep count of how many rotations left
3. Rotate
4. decrement count of rotations
5. convert back to int
6. return


'''
def max_rotation(number):
    string_num = str(number)
    length = len(string_num)
    rotations = 0

    while rotations < length:
        end = string_num[rotations]
        left_side = string_num[:rotations]
        right_side = string_num[rotations + 1:]
        string_num = left_side + right_side + end
        rotations += 1
    
    return int(string_num)

print(max_rotation(735291) == 321579)          # True
print(max_rotation(3) == 3)                    # True
print(max_rotation(35) == 53)                  # True
print(max_rotation(8703529146) == 7321609845)  # True

# Note that the final sequence here is `015`. The leading
# zero gets dropped, though, since we're working with
# an integer.
print(max_rotation(105) == 15)                 # True

File: ls215/smallProblems/med1/3.js

'use strict';

/*
input: Number
output:
rules:

735291 -> 6 -> 352917
012345

352917 -> 5 ->
012345

length = 6
end = 6 - 5 = 1

front = 3
mid = 2917
end = 5

Q:


D:


A:


*/




function maxRotation(number) {
  let strNum = String(number);
  let newNum = number;

  for (let i = strNum.length; i > 0; i -= 1) {
    newNum = rotateRightmostDigits(newNum, i);
  }

  return newNum;
}

function rotateRightmostDigits(number, n) {
  let strNum = String(number);
  let length = strNum.length;
  let end = length - n;
  let front = strNum.slice(0, end);
  let middle = strNum.slice(end + 1);
  let endingNum = strNum[end];


  return Number(front + middle + endingNum);
}



console.log(maxRotation(735291));          // 321579
console.log(maxRotation(3));               // 3
console.log(maxRotation(35));              // 53
console.log(maxRotation(105));             // 15 -- the leading zero gets dropped
console.log(maxRotation(8703529146));      // 7321609845